// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/prompt-version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const constructs_1 = require("constructs");
/**
 * Creates a version of the prompt.
 *
 * Use this to create a static snapshot of your prompt that can be deployed
 * to production. Versions allow you to easily switch between different
 * configurations for your prompt and update your application with the most
 * appropriate version for your use-case.
 * @resource AWS::Bedrock::PromptVersion
 * @see https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-management-deploy.html
 */
class PromptVersion extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.prompt = props.prompt;
        // L1 instantiation
        this._resource = new aws_cdk_lib_1.aws_bedrock.CfnPromptVersion(this, id, {
            description: props.description,
            promptArn: props.prompt.promptArn,
        });
        this.versionArn = this._resource.attrArn;
        this.version = this._resource.attrVersion;
    }
}
exports.PromptVersion = PromptVersion;
_a = JSII_RTTI_SYMBOL_1;
PromptVersion[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.PromptVersion", version: "0.1.264" };
